File: histo_2d.py

import os
import sys


def mkdir(path):
	try:
		os.mkdir(path)
	except:
		print('cannot make '+path)


def system(cmd):
	print('----------------------------------------------------')
	print(' histo.py RUN COMMAND')
	print('', cmd)
	print('----------------------------------------------------')
	rt = os.system(cmd)
	if (rt/8==0):
		print('----------------------------------------------------')
		print(' COMMAND SUCCESSFUL')
		print('----------------------------------------------------')
	else:
		print('----------------------------------------------------')
		print(' COMMAND FAILURE')
		print('----------------------------------------------------')


#----------------------------
# Check command arguments
#----------------------------
if (len(sys.argv)<8):
	print('Usagye')
	print('')
	print('   python3 histo.py dataset model seed minx maxx nbin maxC1')
	print('')
	print('  dataset:   cifar10 cifar100 imagenette2 mnist')
	print('  model:     resnet18 resnet50 vgg19')
	print('  seed:      integer seed for which model to load')
	print('  batchsize: how many images to process at once')
	print('  minx:      minimum cutoff for histogram')
	print('  maxx:      maximum cutoff for histogram')
	print('  nbin:      number of bins for histogram')
	print('  maxC1:     maximum number of channels to create histograms')
	sys.exit(1)

arg_dataset    =         sys.argv[1]
arg_model      =         sys.argv[2]
arg_seed       =     int(sys.argv[3])
arg_minx       =   float(sys.argv[4])
arg_maxx       =   float(sys.argv[5])
arg_nbin       =     int(sys.argv[6])
arg_maxC1      =     int(sys.argv[7])

print('arg_dataset', arg_dataset)
print('arg_model', arg_model)
print('arg_seed', arg_seed)
print('arg_minx', arg_minx)
print('arg_maxx', arg_maxx)
print('arg_nbin',  arg_nbin)
print('arg_maxC1', arg_maxC1)

#
# Construct the folders
#
featdir = 'features/%s_%s_%d' % (arg_dataset,arg_model,arg_seed)
histdir = 'histogram_2d/%s_%s_%d' % (arg_dataset,arg_model,arg_seed)
mkdir('histogram_2d')
mkdir(histdir)


#
# Compile the C program
#
system("make")

#
# For every feature array
#
for arr in ('X_train', 'A_train', 'B_train', 'C_train', 'D_train', 'E_train', 'X_test', 'A_test', 'B_test', 'C_test', 'D_test', 'E_test'):

	#cmd = './histo_kernel %s/%s.bin %s/%s.csv %f %f %d %f' % (featdir,arr,histdir,arr,arg_minx,arg_maxx,arg_nbin,arg_width)
	cmd = './histo_2d %s/%s.bin %s/%s.bin %f %f %d %d' % (featdir,arr,histdir,arr,arg_minx,arg_maxx,arg_nbin,arg_maxC1)
	system(cmd)


print('Done!\n')

File: histo_2d.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "mmap.h"

#define int64 long long int


// Easy string formatting
char path[4096];
char *cat(const char*str1, const char*str2) {
	(sprintf)(path, "%s%s", str1, str2);
	return path;
}

//---------
// Command Arguments
//---------
char *arg_infile;
char *arg_outfile;
float arg_minx;
float arg_maxx;
int   arg_nbin;
int   arg_maxC1;

//---------
// Histogram file
//---------
Binfile datafile;
float *X;
int64 N;
int64 C;
int64 sY;
int64 sX;
int64 CsYsX;
int64 sYsX;


//---------
// Output histogram
//---------
Binfile outhist;
int *histo;

int main(int argc, char**argv)
{
	int64 n,c,c1,c2,y,x,b1,b2;

	//----------------------
	// Read Command Arugments
	//----------------------
	if (argc<7) {
		printf("Usage:\n");
		printf("   ./histo_2d infile outfile minx maxx nbin maxC1\n");
		printf("\n");
		printf("   infile:      in .bin feature file\n");
		printf("   outfile:     output binary file\n");
		printf("   minx,maxx:   x-interval for histogram binning\n");
		printf("   nbin:        how many histogram bins\n");
		printf("\n");
		exit(1);
	}
	arg_infile  =      argv[1];
	arg_outfile =      argv[2];
	arg_minx    = atof(argv[3]);
	arg_maxx    = atof(argv[4]);
	arg_nbin    = atoi(argv[5]);
	arg_maxC1   = atoi(argv[6]);

	printf("arg_infile: %s\n",  arg_infile);
	printf("arg_outfile: %s\n", arg_outfile);
	printf("arg_minx %f\n", arg_minx);
	printf("arg_maxx %f\n", arg_maxx);
	printf("arg_nbin %d\n", arg_nbin);
	printf("arg_maxC1 %d\n", arg_maxC1);

	//----------------------
	// Memory map the data histogram file
	//----------------------
	datafile = MapBinfileR(arg_infile);
	X = (float*)datafile.data;
	N = datafile.shape[0];
	C = datafile.shape[1];
	sY = datafile.shape[2];
	sX = datafile.shape[3];
	CsYsX = C*sY*sX;
	sYsX = sY*sX;

	printf("N %lld C %lld sY %lld sX %lld CsYsX %lld sYsX %lld\n", N, C, sY, sX, CsYsX, sYsX);


	printf("----------------------\n");
	printf(" Memory map the histograms\n");
	printf("----------------------\n");
		
	int outhist_shape[5];
	outhist_shape[0] = arg_maxC1;
	outhist_shape[1] = C;
	outhist_shape[2] = arg_nbin;
	outhist_shape[3] = arg_nbin;
	outhist_shape[4] = 0;
	outhist = MapBinfileRW(arg_outfile, outhist_shape, INT32);
	int64 BB  = arg_nbin*arg_nbin;
	int64 CBB = C*BB;
	int64 CCBB = arg_maxC1*CBB;
	histo = (int*)outhist.data;
	
	printf("fd %d BB %lld CBB %lld CCBB %lld histo %p\n", outhist.fd, BB, CBB, CCBB, histo);
	
	printf("----------------------\n");
	printf(" Zero out the histogram\n");
	printf("----------------------\n");
	int64 i;
	for (i=0; i<CCBB; i++)
		histo[i] = 0;

	printf("----------------------\n");
	printf(" Compute the histograms\n");
	printf("----------------------\n");

	float scaling = (float)arg_nbin / (arg_maxx-arg_minx);

	// For every data element
	for (n=0; n<N; n++) {
		if (n%100==0) printf("n %lld/%lld\n", n, N);
	for (c1=0; c1<arg_maxC1; c1++) {
	for (c2=0; c2<C; c2++) {
	for (y=0; y<sY; y++) {
	for (x=0; x<sX; x++) {

		int64 idx1 = n*CsYsX + c1*sYsX + y*sX + x;
		int64 idx2 = n*CsYsX + c2*sYsX + y*sX + x;

		//-------------------
		// Identify bin 1
		//-------------------

		// Which bin (floating point) ?
		float val1 = X[idx1];
		float fbin1 = (val1-arg_minx) * scaling;
		if (fbin1<0) {
			//printf("WARNING fbin1 out of bounds val1 %f fbin1 %f\n", val1, fbin1);
			continue;
		}
		else if (fbin1>=arg_nbin) {
			//printf("WARNING fbin1 out of bounds val1 %f fbin1 %f\n", val1, fbin1);
			continue;
		}

		// Which bin (integer)
		int ibin1 = (int)fbin1;
		if (ibin1<0)
			ibin1=0;
		if (ibin1>arg_nbin-1)
			ibin1=arg_nbin-1;

		//-------------------
		// Identify bin 2
		//-------------------

		// Which bin (floating point) ?
		float val2 = X[idx2];
		float fbin2 = (val2-arg_minx) * scaling;
		if (fbin2<0) {
			//printf("WARNING fbin2 out of bounds val2 %f fbin2 %f\n", val2, fbin2);
			continue;
		}
		else if (fbin2>=arg_nbin) {
			//printf("WARNING fbin2 out of bounds val2 %f fbin2 %f\n", val2, fbin2);
			continue;
		}

		// Which bin (integer)
		int ibin2 = (int)fbin2;
		if (ibin2<0)
			ibin2=0;
		if (ibin2>arg_nbin-2)
			ibin2=arg_nbin-2;

		// Bin the value
		i64 outidx = c1*CBB + c2*BB + ibin1*arg_nbin + ibin2;
		histo[outidx]++;

	}}}}}


	printf("----------------------\n");
	printf(" Unmap the dataset features\n");
	printf("----------------------\n");
	UnmapBinfile(datafile);

	printf("----------------------\n");
	printf(" Unmap the output histogram\n");
	printf("----------------------\n");
	UnmapBinfile(outhist);

	printf("Success\n");

	return 0;
}
